### ISSUE TITLE
Deny the `as` casts and route the rest through SaturatingInto/ApproxInto

### ISSUE THREAD

**Issue by kozalosev** [kozalosev/DickGrowerBot#161]
Follow-up to the commit that added `SaturatingInto` and `ApproxInto` (`domain_types::traits`).

That commit gave the conversions a name and moved the domain-type call sites onto them. It did
**not** stop anyone writing `as`, so the rule is still something to remember rather than something
the build enforces. This issue is the other half.

## What this buys

A bare `as` on a number stops compiling. The same shape `clippy.toml`'s `disallowed-methods`
already gives `literal!`: the rule is enforced instead of remembered.

It also settles the one open question about `ApproxInto`. Today it changes no behaviour — the float
directions of `as` already do the right thing (RFC 2484), so those impls *are* `as` with a name on
them. Their justification is entirely downstream of this issue. Without the denial they are dead
weight; with it they are what keeps a denied cast from being holed by `#[allow]`. See
`src/handlers/perks.rs` before the parent commit:

```rust
let payout = payout_coefficient.scale(base_increment as f64).round() as i64;
```

Two casts on one line, one harmless and one not. With the denial and no vocabulary for the harmless
one, the only way to compile it is an `#[allow]` on the whole statement — which then silences every
cast added to that line later, including a genuinely wrapping one. With both traits the line is
total and needs no allow. **If this issue is closed as won't-do, `ApproxInto` should be dropped.**

## The lints

A new section in the root `Cargo.toml`:

```toml
[lints.clippy]
cast_possible_wrap = "deny"        # u64 -> i64
cast_possible_truncation = "deny"  # u64 -> u32, f64 -> i64
cast_sign_loss = "deny"            # i64 -> u64
```

`domain_types` needs its own section, or an `#[allow]` where the primitive impls deliberately use
`as`; they already carry one for the clippy lints by name.

## The remaining sites

All plain primitives rather than domain types, which is why they were left out. The primitive impls
added by the parent commit already serve them, so this part is mechanical:

- [ ] `src/metrics.rs` — `self.pool.size() as i64`, `self.pool.num_idle() as i64`
- [ ] `src/scheduler/shrink.rs` — `events.len() as u64` and two more of the same shape
- [ ] `src/scheduler/deletions.rs` — `due.len() as f64`
- [ ] `src/handlers/dick.rs` — `(i+1) as i64`
- [ ] `src/users/mod.rs` — `uid.0 as i64`, teloxide's `UserId` rather than ours
- [ ] `src/domain/primitives/pagination.rs` — the two `From` impls, `value as i32` and `value as u16`

## The ones that want an allow rather than a conversion

- [ ] `src/config/self_destruction.rs` — `1 << (group as u8)`, a discriminant, not a number
- [ ] anywhere a cast is provably exact — `#[allow]` is the right answer, as long as it carries the
      reason and sits on the narrowest scope that works

## Done when

`cargo clippy --all-targets -- -D warnings` passes with the three lints denied, and no `#[allow]`
in the tree covers a statement wider than the single cast it exists for.

**Comment by kozalosev** [comment]
Done on `feature/self-destruction-persistence`.

Five lints rather than the three planned. `cast_precision_loss` cost one call site and is what enforces `ApproxInto` on the int-to-float direction; `cast_lossless` cost six and is the one that makes an exact conversion say `From`, which is the row of the table the issue didn't cover.

Two things came out differently from the plan:

* **`1 << (group as u8)` needs no `#[allow]`.** No lint fires on it — casting a fieldless enum to an integer is `cast_enum_truncation`'s business, and that one isn't denied. Same for `chat_id.0 as i64`, which is sqlx's type-override syntax rather than a cast, and `UID as u64` inside `literal!`, which the build evaluates.
* **The macro's generated code was a hole.** `generate_division_operator_impls` emits `self.0 as f64`, and a proc macro's spans are exempt from clippy, so the denial never reached it. Now routed through `ApproxInto`. That doesn't fix the precision loss the TODO there describes (an f64 holds integers exactly only to 2^53), but it stops it being invisible.

The convention is written up in CLAUDE.md next to the domain-wrapper rule.
